### PR TITLE
dynamo: support placement kwargs for DTensor.to_local()

### PR DESCRIPTION
**Pull request by bdhirsh**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118802
* #118670
* #118669
* __->__ #119947
* #118803

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @aakhundov

### CODE REVIEW

File: torch/_dynamo/variables/tensor.py [hunk 1]
```diff
@@ -807,6 +807,26 @@ def _register_hook_trampoline(tensor):
                 return tensor
 
             from .builder import wrap_fx_proxy
+            # TODO: figure out if this is necessary
+            if name =='to_local':
```
Comment (wanchaol) [inline torch/_dynamo/variables/tensor.py:811]: wondering can we unify the logic with `from_local`? https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/variables/torch.py#L525

We can possibly see if we can make `to_local` be an TorchInGraphFunctionVariable too

Comment (bdhirsh) [reply torch/_dynamo/variables/tensor.py]: Turns out that change wasn't actually necessary, so I ended up removing it.

I add a different change though to handle hooks (see the PR update) - let me know if you think it's too boilerplate-y and worth refactoring with the to_local code!

File: torch/_dynamo/variables/tensor.py [hunk 2]
```diff
@@ -761,6 +761,32 @@ def redistribute_fn_with_prim_types(x):
             ),
         )
 
+    def method_to_local(self, *args, **kwargs):
```
[the remark this answers is not in the document]
Comment (wanchaol) [reply torch/_dynamo/variables/tensor.py:764]: IMO for (1) that's actually the only reason I think haha. The best trial we can do after this is to allow registering custom types that does not care about serialization (a lot of them don't I think), allowing dynamo to trace and leave custom types in the graph according to user intention is sth I think legit to do. cc @ezyang as we discussed this before

**Review by wanchaol — APPROVED** [review]
I think this looks reasonable to me

**Review by yoyoyocmu — APPROVED** [review]

**Pull request closed without merging**